### PR TITLE
[SPARK-44435][SPARK-44484][3.5][SS][CONNECT] Tests for foreachBatch and Listener

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Add several new test cases for streaming foreachBatch and streaming query listener events to test various scenarios.
Also merge this PR to 3.5 to fix the test error: https://github.com/apache/spark/pull/42077

### Why are the changes needed?

More tests is better

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Test only change

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 1]
```diff
@@ -181,6 +181,7 @@ class StreamingQueryProgress private[spark] (
       ("name" -> JString(name)) ~
       ("timestamp" -> JString(timestamp)) ~
       ("batchId" -> JInt(batchId)) ~
+      ("batchDuration" -> JInt(batchDuration)) ~
```
Comment (dongjoon-hyun) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:184]: Thank you for making a new PR.

However, given the previous discussion on SPARK-44484. We cannot backport a new feature like this under a different `Tests for foreachBatch and Listener` title.

- https://github.com/apache/spark/pull/42077

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
If SPARK-44484 is inevitable for SPARK-44435, I'd prefer to stop backporting because we are in RC2 stage already, @WweiL . 

Screenshot 2023-08-24 at 3 28 54 PM

**Comment by WweiL** [comment]
@dongjoon-hyun I see. I'll close this. Thanks for the context.

**Comment by dongjoon-hyun** [comment]
Thank you for closing this, @WweiL .

File: python/pyspark/sql/streaming/listener.py [hunk 2]
```diff
@@ -477,7 +477,8 @@ def fromJson(cls, j: Dict[str, Any]) -> "StreamingQueryProgress":
             name=j["name"],
             timestamp=j["timestamp"],
             batchId=j["batchId"],
-            batchDuration=j["batchDuration"],
+            # before spark 4.0, batchDuration is not in the json method of jvm side StreamingQueryProgress
+            batchDuration=j["batchDuration"] if "batchDuration" in j else None,
```
Comment (WweiL) [inline python/pyspark/sql/streaming/listener.py:481]: @dongjoon-hyun Hi Dongjoon, sorry for the back and forth. On second thought I actually find out that the newly added tests and the test failure in https://github.com/apache/spark/pull/42521#issuecomment-1691547730 actually finds out a bug. Here before I assume `batchDuration` is always in the passed in json, but before 4.0 it is not there.

Given that we don't add the change in https://github.com/apache/spark/pull/42077, this check is needed. I reverted that commit, and add this check.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
It's totally fine. :)

If this is a bug fix, please make a new JIRA to proceed it instead of SPARK-44435 or SPARK-44484. The JIRA title should be clear that is a bug fix. We can merge that first.

**Comment by WweiL** [comment]
fixed in https://github.com/apache/spark/commit/7be69bf7da036282c2c7c0b62c32e7666fa1b579

**Pull request closed without merging**